src/main/java/jenkins/plugins/git/GitRepoSizeEstimator.java

```diff
@@ -0,0 +1,168 @@
+package jenkins.plugins.git;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.EnvVars;
+import hudson.ExtensionList;
+import hudson.ExtensionPoint;
+import hudson.model.TaskListener;
+import hudson.plugins.git.GitTool;
+import hudson.plugins.git.util.GitUtils;
+import jenkins.model.Jenkins;
+import org.apache.commons.io.FileUtils;
+import org.jenkinsci.plugins.gitclient.Git;
+import org.jenkinsci.plugins.gitclient.GitClient;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * A class which allows {@link AbstractGitSCMSource} to estimate the size of a repository from a distance
+ * without requiring a local checkout.
+ */
+public class GitRepoSizeEstimator {
+
+    private long sizeOfRepo = 0L;
+    private String implementation;
+    private String gitTool;
+    public static final int SIZE_TO_SWITCH = 50000;
+
+    /**
+     * Instantiate class using {@link AbstractGitSCMSource}. It looks for a cached .git directory first, calculates the
+     * size if it is found else checks if the extension point has been implemented and asks for the size.
+     * @param source the {@link AbstractGitSCMSource}
+     * @throws IOException
+     * @throws InterruptedException
+     */
+    public GitRepoSizeEstimator(@NonNull AbstractGitSCMSource source) throws IOException, InterruptedException {
+        boolean useCache;
+        boolean useAPI = false;
+
+        useCache = setSizeFromCache(source);
+
+        if (useCache) {
+            implementation = determineSwitchOnSize(sizeOfRepo);
+        } else {
+            useAPI = setSizeFromAPI(source.getRemote());
+        }
+
+        if (useAPI) {
+            implementation = determineSwitchOnSize(sizeOfRepo);
+        }
+
+        if (!useAPI && !useCache) {
+            implementation = "DEFAULT";
+        }
+        determineGitTool(implementation);
```

Comment:
That concerns me. You're mixing the Extension and the Estimator. I think you should decouple them doing something like:
- using the URL there is a mechanism to detect the correct Extension
- Retrieve that concrete extension
- Invoke a method of the Extension retrieving the size
- Store the retrieved value in the field

Right now, your code:
- From the Estimator it is executing a method in the Extension
- From the Extension it is updating a field in the Estimator
There is a cycle, a circular dependency between the two classes

Reply:
This was concerning me too. Was not sure whether it has to be raised or not. Need separation of concern here.